source/auth_credentials.c

```diff
@@ -544,3 +548,124 @@ PyObject *aws_py_credentials_provider_new_chain(PyObject *self, PyObject *args)
     Py_XDECREF(capsule);
     return NULL;
 }
+
+static int s_credentials_provider_py_provider_get_credentials(
+    struct aws_credentials_provider *provider,
+    aws_on_get_credentials_callback_fn callback,
+    void *user_data) {
+    struct aws_allocator *allocator = provider->allocator;
+
+    struct aws_string *access_key_id = NULL;
+    struct aws_string *secret_access_key = NULL;
+    struct aws_string *session_token = NULL;
+    struct aws_credentials *credentials = NULL;
+    int error_code = AWS_ERROR_SUCCESS;
+    struct credentials_provider_binding *binding = provider->impl;
+
+    PyGILState_STATE state;
+    if (aws_py_gilstate_ensure(&state)) {
+        return AWS_OP_ERR; /* Python has shut down. Nothing matters anymore, but don't crash */
+    }
+
+    PyObject *dict = PyObject_CallMethod(binding->py_provider, "get_credential", "()");
+
+    PyObject *py_key = PyDict_GetItemString(dict, "AccessKeyId");
+    PyObject *py_secret_key = PyDict_GetItemString(dict, "SecretAccessKey");
+    PyObject *py_token = PyDict_GetItemString(dict, "SessionToken");
+    PyObject *py_expiration = PyDict_GetItemString(dict, "Expiration");
+
+    struct aws_byte_cursor cursor;
+    if (py_key) {
+        cursor = aws_byte_cursor_from_pyunicode(py_key);
+        access_key_id = aws_string_new_from_cursor(allocator, &cursor);
+    }
+    if (py_secret_key) {
+        cursor = aws_byte_cursor_from_pyunicode(py_secret_key);
+        secret_access_key = aws_string_new_from_cursor(allocator, &cursor);
+    }
+    if (py_token) {
+        cursor = aws_byte_cursor_from_pyunicode(py_token);
+        session_token = aws_string_new_from_cursor(allocator, &cursor);
+    }
+    uint64_t expiration = UINT64_MAX;
+    if (py_expiration) {
+        expiration = PyLong_AsUnsignedLongLong(py_expiration);
+    }
+    PyGILState_Release(state);
+    if (access_key_id != NULL && secret_access_key != NULL) {
+        credentials =
+            aws_credentials_new_from_string(allocator, access_key_id, secret_access_key, session_token, expiration);
+        if (credentials == NULL) {
+            error_code = aws_last_error();
+        }
+    } else {
+        error_code = AWS_AUTH_CREDENTIALS_PROVIDER_INVALID_REVERT;
+    }
+
+    callback(credentials, error_code, user_data);
+
+    aws_credentials_release(credentials);
+    aws_string_destroy(session_token);
+    aws_string_destroy(secret_access_key);
+    aws_string_destroy(access_key_id);
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_credentials_provider_py_provider_destroy(struct aws_credentials_provider *provider) {
+    if (provider && provider->shutdown_options.shutdown_callback) {
```

Comment:
I would switch these statements safely (save off the callback and user data in locals).  It's possible for tests to be gated on this completing and if they do a mem check before the release completes, the mem check will fail.

Reply:
did not get what you mean by "save off the callback and user data in locals" here...

Reply:
if you release the memory you can't use it anymore so values from it need to be pulled off first.